File: mobile/sme_mobile/lib/services/payment_flow.dart
import '../models/billing_models.dart';

/// The payment flow's decisions, separate from its widgets so they can be
/// unit-tested: which gateway a method goes through, and what a typed
/// amount means. Mirrors frontend/src/features/billing/paymentFlow.ts.
enum PayMethod { card, qr, cash }

extension PayMethodX on PayMethod {
  String get apiName => switch (this) { PayMethod.card => 'Card', PayMethod.qr => 'QR', PayMethod.cash => 'Cash' };
  String get title => switch (this) { PayMethod.card => 'Card', PayMethod.qr => 'QR / wallet', PayMethod.cash => 'Cash' };

  static PayMethod fromSaved(String type) => switch (type) {
        'QR' || 'PayPal' => PayMethod.qr,
        'Cash' => PayMethod.cash,
        _ => PayMethod.card,
      };
}

/// Cash never touches a gateway. Card goes to Stripe (its hosted page) when
/// the business has it, QR to PayPal when set up; otherwise the sandbox.
String? providerFor(PayMethod method, AvailableProviders? available) {
  if (method == PayMethod.cash) return null;
  final providers = available?.providers ?? const <ProviderOption>[];
  if (method == PayMethod.card) {
    if (available?.has('Stripe') ?? false) return 'Stripe';
    return providers.isNotEmpty ? providers.first.provider : 'Manual';
  }
  return (available?.has('PayPal') ?? false) ? 'PayPal' : 'Manual';
}

/// Null when the typed amount is fine (blank means "the full balance").
String? validatePayAmount(String raw, double balance) {
  final text = raw.trim();
  if (text.isEmpty) return null;
  final value = double.tryParse(text);
  if (value == null || value <= 0) return 'Enter an amount greater than 0.';
  if ((value * 100).round() > (balance * 100).round()) return 'You can pay at most ${balance.toStringAsFixed(2)}.';
  return null;
}

double amountToPay(String raw, double balance) => raw.trim().isEmpty ? balance : double.parse(raw.trim());

/// What the customer is told after the provider answers.
String resultMessage(String status) => switch (status) {
      'Succeeded' => 'Payment received - your receipt is ready.',
      'Failed' => 'The payment did not go through. No money was taken.',
      _ => 'The payment is still processing. It will update once the provider confirms it.',
    };

File: frontend/src/features/billing/paymentFlow.ts
import type { AvailableProviders, CheckoutResponse } from './billingApi';

/* The payment flow's state machine, separate from its UI so the steps can
 * be tested: choose method -> (start checkout) -> provider step -> result. */

export type PayMethod = 'Card' | 'QR' | 'Cash';

export type PaymentStep =
  | { kind: 'choose' }
  | { kind: 'starting' }
  | { kind: 'sandbox'; session: CheckoutResponse }
  | { kind: 'stripe'; session: CheckoutResponse }
  | { kind: 'paypal'; session: CheckoutResponse }
  | { kind: 'cash' }
  | { kind: 'confirming'; session: CheckoutResponse }
  | { kind: 'done'; status: 'Succeeded' | 'Failed' | 'Pending'; message: string }
  | { kind: 'error'; message: string };

/** Which provider a method goes through. Cash never touches a gateway -
 *  it is paid at the counter and recorded by staff. */
export function providerFor(method: PayMethod, available: AvailableProviders | null): 'Stripe' | 'PayPal' | 'Manual' | null {
  if (method === 'Cash') return null;
  const providers = available?.providers ?? [];
  if (method === 'Card') return providers.find((p) => p.provider === 'Stripe')?.provider ?? providers[0]?.provider ?? 'Manual';
  // QR payments (LankaQR and similar) settle through the business's own
  // acquirer; online we run them through the sandbox unless PayPal is set up.
  return providers.find((p) => p.provider === 'PayPal')?.provider ?? 'Manual';
}

/** After checkout starts, where the flow goes next. */
export function stepForSession(session: CheckoutResponse): PaymentStep {
  if (session.status === 'Succeeded') return { kind: 'done', status: 'Succeeded', message: 'Payment received.' };
  if (session.status === 'Failed') return { kind: 'done', status: 'Failed', message: 'The payment could not be started.' };
  if (session.provider === 'Stripe' && session.clientSecret) return { kind: 'stripe', session };
  if (session.provider === 'PayPal' && session.redirectUrl) return { kind: 'paypal', session };
  return { kind: 'sandbox', session };
}

export function resultStep(status: string): PaymentStep {
  if (status === 'Succeeded') return { kind: 'done', status: 'Succeeded', message: 'Payment received - your receipt is ready.' };
  if (status === 'Failed') return { kind: 'done', status: 'Failed', message: 'The payment did not go through. No money was taken.' };
  return { kind: 'done', status: 'Pending', message: 'The payment is still processing. It will update once the provider confirms it.' };
}

/** Validates a part-payment amount typed by the payer. */
export function validateAmount(raw: string, balance: number): string | null {
  if (raw.trim() === '') return null; // blank = pay the full balance
  const value = Number(raw);
  if (!Number.isFinite(value) || value <= 0) return 'Enter an amount greater than 0.';
  if (Math.round(value * 100) > Math.round(balance * 100)) return `You can pay at most ${balance.toFixed(2)}.`;
  return null;
}

/** Loads Stripe.js once, from Stripe's CDN (it must not be self-hosted). */
let stripeScript: Promise<void> | null = null;
export function loadStripeJs(): Promise<void> {
  if ((window as unknown as { Stripe?: unknown }).Stripe) return Promise.resolve();
  stripeScript ??= new Promise<void>((resolve, reject) => {
    const s = document.createElement('script');
    s.src = 'https://js.stripe.com/v3/';
    s.async = true;
    s.onload = () => resolve();
    s.onerror = () => {
      stripeScript = null;
      reject(new Error('Could not load Stripe.'));
    };
    document.head.appendChild(s);
  });
  return stripeScript;
}
